hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobExecutionRecord.java

```diff
@@ -0,0 +1,307 @@
+            this.endTime = Clock.currentTimeMillis();
+            this.dataMapIndex = ongoingDataMapIndex();
+        } else {
+            // we don't update the other fields because they only pertain to a successful snapshot
+        }
+    }
+
+    public long snapshotId() {
+        return this.snapshotId;
+    }
+
+    public int dataMapIndex() {
+        return this.dataMapIndex;
+    }
+
+    /**
+     * Returns the index of the data map into which the new snapshot will be
+     * written.
+     */
+    int ongoingDataMapIndex() {
+        assert this.dataMapIndex == 0 // we'll return 1
+                || this.dataMapIndex == 1 // we'll return 0
+                || this.dataMapIndex == -1 // we'll return 0
+                : "dataMapIndex=" + this.dataMapIndex;
+        return (this.dataMapIndex + 1) & 1;
+    }
+
+    public long startTime() {
+        return this.startTime;
+    }
+
+    public long endTime() {
+        return this.endTime;
+    }
+
+    public long duration() {
+        return this.endTime - this.startTime;
+    }
+
+    /**
+     * Net number of bytes in primary copy. Doesn't include IMap overhead and backup copies.
+     */
+    public long numBytes() {
+        return this.numBytes;
+    }
+
+    /**
+     * Number of snapshot keys (after exploding chunks).
+     */
+    public long numKeys() {
+        return this.numKeys;
+    }
+
+    /**
+     * Number of chunks the snapshot is stored in. One chunk is one IMap entry,
+     * so this is the number of entries in the data map.
+     */
+    public long numChunks() {
+        return this.numChunks;
+    }
+
+    public long ongoingSnapshotId() {
+        return this.ongoingSnapshotId;
+    }
+
+    public long ongoingSnapshotStartTime() {
+        return this.ongoingSnapshotStartTime;
+    }
+
+    @Nullable
+    public String lastFailureText() {
+        return this.lastFailureText;
+    }
+
+    long getTimestamp() {
+        return timestamp.get();
+    }
+
+    /**
+     * Sets the timestamp to:
+     *   <pre>max(Clock.currentTimeMillis(), this.timestamp + 1);</pre>
+     * <p>
+     * In other words, after this call the timestamp is guaranteed to be
+     * incremented by at least 1 and be no smaller than the current wall clock
+     * time.
+     */
+    void updateTimestamp() {
+        timestamp.updateAndGet(v -> Math.max(Clock.currentTimeMillis(), v + 1));
+    }
+
+    public String successfulSnapshotDataMapName(long jobId) {
+        if (snapshotId() < 0) {
+            throw new IllegalStateException("No successful snapshot");
+        }
+        return snapshotDataMapName(jobId, dataMapIndex());
+    }
+
+    @Override
+    public int getFactoryId() {
+        return JetInitDataSerializerHook.FACTORY_ID;
+    }
+
+    @Override
+    public int getId() {
+        return JetInitDataSerializerHook.JOB_EXECUTION_RECORD;
+    }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
```

Comment:
would it be a good idea to have a unit test for this, since a lot of fields are read/written?

Reply:
added